crypto/init.c

```diff
@@ -78,13 +78,137 @@ DEFINE_RUN_ONCE_STATIC(ossl_init_base)
     return 0;
 }
 
+/*
+ * Store of provider handles for which there OPENSSL_cleanup() is set up as
+ * and exit function.  The handle NULL represents the application.
+ *
+ * This allows several components (application and providers alike) to share
+ * the same instance of libcrypto, and to ensure that OPENSSL_cleanup() isn't
+ * called too early.
+ *
+ * This array has a fixed size for safety reasons.
+ * OPENSSL_realloc() could have been used for a more dynamic array, but that
+ * also comes with a risk of sudden and surprising failure.
+ * Our stack implementation shouldn't be used either, since it calls
+ * ERR_raise(), which might recurse back into OPENSSL_init_crypto().
+ */
+static const OSSL_CORE_HANDLE *handles[100] = { NULL, };
+static size_t handles_num = 0;
+
+static int ossl_init_register_cleanup(const OSSL_CORE_HANDLE *handle)
+{
+    size_t i;
+    int ret;
+
+    if (!CRYPTO_THREAD_write_lock(init_lock))
+        return -1;
+
+#ifdef OPENSSL_INIT_DEBUG
+    fprintf(stderr,
+            "OPENSSL_INIT: ossl_init_register_cleanup(handle = %p)\n",
+            (void *)handle);
+#endif
+
+    /*
+     * Check if this handle is already registered, as we only allow
+     * one entry each...  Finding one is success.
+     */
+    for (i = 0; i < handles_num; i++)
+        if (handles[i] == handle)
+            break;
+
+    ret = i;
+    if (i < handles_num)
+        goto end;
+
+    ret = -1;
+    if (handles_num < OSSL_NELEM(handles)) {
```

Comment:
What about making this an ossl_assert?
It will make finding an overflow a lot easier.

Reply:
Sure

Reply:
Done